### DISCUSSION TITLE
Using TFT_eSPI with ESPHome to drive a "3.5inch RPi display"

### DISCUSSION THREAD

**Discussion by Gyosa3** [Bodmer/TFT_eSPI#2868] [General]
Hello, I bought some time ago a 3.5inch display for a Raspberry Pi Zero to display a mini dashboard with Home Assistant values (via Chromium). This setup is too slow and not responsive and chromium is now even unsupported on this Pi.

It looks like this:
![image](https://github.com/Bodmer/TFT_eSPI/assets/51777889/c1661141-b5b3-498d-a323-92075755f7b5)

So I wanted to reuse the display with an ESP32 board and native graphics to make it better. I first tried with ESPHome and the ILI9xxx component but it did not work although the display is an ILI9486.

I then turned to the Arduino IDE and this library, and found out that the display works well when selecting the following options in the  `User_setup.h` file:
```
#define RPI_DISPLAY_TYPE
#define RPI_ILI9486_DRIVER
```
This display does not work with the generic ILI9486 driver as found in this lib and in the the ILI9xxx component of ESPHome, but rather with a specific one only available in this lib. Touch screen also works well with the right conf.

I therefore went back to ESPHome and tried to use this library instead of the ILI9xxx component, so that I could select the proper driver for use. I am nearly there and the ESPHome definition looks like:
```
esphome:
  name: esphome-web-c7a87c
  friendly_name: ESPHome Web c7a87c
  libraries:
    - SPI
    - FS
    - SPIFFS
    - TFT_eSPI
  includes:
    - RPi35Display.h
```
The file `RPi35Display.h` contains a constructor for a TFT_eSPI object as on the Arduino IDE. This constructor is later called in the lambda section of a custom component in ESPHome. Not obvious but finally everything compiles and it "works".

However the library is downloaded and compiled in a place that I cannot identify, and therefore I do not have access to the `User_setup.h` file and therefore the display does not work properly because I cannot select the proper driver and other items.

I would therefore like to ask if anyone has already embedded this library in ESPHome to instantiate a custom display, and if yes, where and how would you access the user setup so that the right parameters are passed to the library? I have skimmed all the forums that I could search and I could not find any example of using this lib in ESPHome in a custom component.

The only remaining option that I would see is to manually copy and include the whole file tree of the library as if that was custom code and not a library. So not in the "libraries:" section but in the "includes:" section of ESPHome. And loose all library management...

Any other idea?

Thanks a lot :)

**Comment by Gyosa3** [comment]
for completeness I have also tried to force the platformio.ini file to include configuration items but I could not manage to succeed with it.

I tried different options like this one:
```
platformio_options:
   build_flags: "-DUSER_SETUP_LOADED=1 -DRPI_ILI9486_DRIVER -DTOUCH_CS=21"
```
And these flags are not reflected in the autogenerated platformio.ini file.

![image](https://github.com/Bodmer/TFT_eSPI/assets/51777889/881e80fa-8795-4ef7-a4bd-e02e01234d49)

**Comment by Gyosa3** [comment]
**Update:** As anticipated in my first post I have managed to compile the lib with the proper driver by copying the library locally on the Pi, changing the config file and link the library locally in ESPHOME with:
```
esphome:
  name: esphome-web-c7a87c
  friendly_name: ESPHome Web c7a87c
  libraries:
    - SPI
    - FS
    - SPIFFS
    - file:///config/esphome/tft_espi
  includes:
    - RPi35Display.h
```
The lib dependency is changed to the local one
![image](https://github.com/Bodmer/TFT_eSPI/assets/51777889/3b4ea07d-95c0-4e9a-891f-a65512b5dc09)
and it compiles well, this time with the right driver and GPIO config.
Only at the cost of having to maintain the library manually on the filesystem, but well, if there's no other way...

or is there? thanks for any feedback or opinion on this matter!

**Reply by fsadak** [reply]
Hi @Gyosa3,
Can you give more information about this process? I have same display and cannot run it except touch function. I tried your code lines but getting file not found error in - file:///config/esphome/tft_espi line. Also where is the RPi35Display.h file?

Thanks in advance...

**Comment by booroondook-eng** [comment]
Please tell us how you described your display in the .yaml file? What platform are you using etc?
Could you please provide us with the ESPHome configuration file for your display?
